proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,495 @@
+  finance developer users from performing operations against the clusters of HR developer users who log into the finance
+  webapp. This could be a serious security problem after the planned multiple identity providers feature is implemented
+  to allow for more differentiation between the users of the two FederationDomains. This problem is compounded by the
+  fact that many upstream OIDC providers use browser cookies to avoiding asking an active user to interactively log in
+  again, and also by the fact that we decided to punt implementing a user consent UI screen. Together, these imply that
+  an attacker from the finance department cluster which runs the client webapp would only need to trick an HR user into
+  clicking on a single link in their web browser or email client for the attacker to be able to gain access to the HR
+  clusters using the identity of the HR user, with no further interaction required by the HR user beyond just clicking
+  on the link.
+
+#### Upgrades
+
+The proposed CRD will be new, so there aren't any upgrade concerns for it. Potential changes to the FederationDomain
+resource are also new additions to an existing resource, so there are again no upgrade concerns there.
+
+The `pinniped-cli` client needs to continue to act as before for backwards compatibility with existing installations of
+the Pinniped CLI on user's machines. Therefore, it should be excluded from any new scope-based requirements which would
+restrict the group memberships from being returned. This will allow mixing of old CLIs with new Supervisors, and new
+CLIs with old Supervisors, in regard to the new Supervisor features proposed herein.
+
+#### Tests
+
+As usual, unit tests should be added for all new/changed code.
+
+Integration tests should be added to mimic the usage patterns of a webapp. Dynamic clients should be configured with
+various options to ensure that the options work as expected. Those clients should be used to perform the authcode flow,
+RFC8693 token exchanges, and TokenCredentialRequest calls. Negative tests should include validation failures on the new
+CRs, and failures to perform actions that are supposed to be disallowed on the client by its configuration.
+
+#### New Dependencies
+
+None are foreseen.
+
+#### Performance Considerations
+
+Some considerations were mentioned previously for client secret option 2 above. No other performance impact is foreseen.
+
+#### Observability Considerations
+
+None are foreseen, aside from the usual error messages and log statements for the new features similar to what is
+already implemented in the Supervisor.
+
+#### Security Considerations
+
+Some security considerations were already mentioned above. Here are a couple more.
+
+##### Ensuring reasonable client secrets
+
+During a login flow, when the client secret is presented in plaintext to the Supervisor’s token endpoint, it could
```

Comment:
Would this validation print some warning to the logs, or prevent logins?

Reply:
I would expect it to prevent logins in order to get the attention of the admin. It could also put a nice message in the logs to help the admin debug.

Reply:
Makes sense to me.